dopple/dopple.py

```diff
@@ -106,63 +110,74 @@ def check_connection(self, timeout):
             time.sleep(SLEEPTIME)
             wait_time += SLEEPTIME
             if wait_time > timeout:
-                raise last_exception if last_exception else TimeoutError
+                if last_exception is not None:
+                    raise last_exception
+                else:
+                    raise TimeoutError()
 
-    def recv(self, max_length):
+    def recv(self, max_length: int) -> Any:
         return self.socket().recv(max_length)
 
-    def sendall(self, data):
+    def sendall(self, data: bytes) -> None:
         try:
-            return self.socket().sendall(data)
+            return self.socket().sendall(data)  # type: ignore
         except OSError as ex:
             if ex.errno == errno.EPIPE:
                 # The connection was terminated by the backend. Try reconnect.
                 self.close()
-                return self.socket().sendall(data)
+                return self.socket().sendall(data)  # type: ignore
             else:
                 raise
 
 
 class NamedPipeConnector(object):
     """Windows named pipe simulating socket."""
 
-    def __init__(self, ipc_path):
+    def __init__(self, ipc_path: str) -> None:
```

Comment:
I intentionally tried to not perform any refactoring at this stage and just focus on getting the code typed and formatted so that it passes CI. Then, I like to add integration tests to ensure any coming refactoring won't break existing functionality.